mozilla-central-patches/aboutnewtabservice.patch

```diff
@@ -0,0 +1,226 @@
++          Services.obs.notifyObservers(null, "newtab-url-changed", ABOUT_URL);
+         }
+         break;
+       case "quit-application-granted":
+-        Services.obs.removeObserver(this, "quit-application-granted");
+-        Services.prefs.removeObserver(PREF_ACTIVITY_STREAM_ENABLED, this);
+-        Services.prefs.removeObserver(PREF_ACTIVITY_STREAM_PRERENDER_ENABLED, this);
++        this.uninit();
+         if (IS_MAIN_PROCESS) {
+           AboutNewTab.uninit();
+@@ -138,4 +151,7 @@ AboutNewTabService.prototype = {
+     }
+     this._activityStreamPrerender = Services.prefs.getBoolPref(PREF_ACTIVITY_STREAM_PRERENDER_ENABLED);
++    if (!IS_RELEASE_OR_BETA) {
++      this._activityStreamDebug = Services.prefs.getBoolPref(PREF_ACTIVITY_STREAM_DEBUG, false);
++    }
+     this._newtabURL = ABOUT_URL;
+     return true;
+@@ -152,5 +168,5 @@ AboutNewTabService.prototype = {
+   get defaultURL() {
+     if (this.activityStreamEnabled) {
+-      return this.activityStreamPrerender ? this.activityStreamPrerenderURL : this.activityStreamURL;
++      return `${BASE_AS_PATH}${this.activityStreamPrerender ? "-prerendered" : ""}${this.activityStreamDebug ? "-debug" : ""}.html`
+     }
+     return LOCAL_NEWTAB_URL;
+@@ -189,10 +205,6 @@ AboutNewTabService.prototype = {
+   },
+ 
+-  get activityStreamURL() {
+-    return ACTIVITY_STREAM_URL;
+-  },
+-
+-  get activityStreamPrerenderURL() {
+-    return ACTIVITY_STREAM_PRERENDER_URL;
++  get activityStreamDebug() {
++    return this._activityStreamDebug;
+   },
+ 
+@@ -202,4 +214,17 @@ AboutNewTabService.prototype = {
+     this.toggleActivityStream(undefined, true);
+     Services.obs.notifyObservers(null, "newtab-url-changed", this._newTabURL);
++  },
++
++  uninit() {
++    if (!this.initialized) {
++      return;
++    }
++    Services.obs.removeObserver(this, "quit-application-granted");
++    Services.prefs.removeObserver(PREF_ACTIVITY_STREAM_ENABLED, this);
++    Services.prefs.removeObserver(PREF_ACTIVITY_STREAM_PRERENDER_ENABLED, this);
++    if (!IS_RELEASE_OR_BETA) {
++      Services.prefs.removeObserver(PREF_ACTIVITY_STREAM_DEBUG, this);
++    }
++    this.initialized = false;
+   }
+ };
+diff --git a/browser/components/newtab/nsIAboutNewTabService.idl b/browser/components/newtab/nsIAboutNewTabService.idl
+--- a/browser/components/newtab/nsIAboutNewTabService.idl
++++ b/browser/components/newtab/nsIAboutNewTabService.idl
+@@ -41,12 +41,7 @@ interface nsIAboutNewTabService : nsISup
+ 
+   /**
+-   * Returns the activity stream resource URL for the newtab page
++   * Returns true if the the debug pref for activity stream is true
+    */
+-  readonly attribute ACString activityStreamURL;
+-
+-  /**
+-   * Returns the prerendered activity stream resource URL for the newtab page
+-   */
+-  readonly attribute ACString activityStreamPrerenderURL;
++  readonly attribute bool activityStreamDebug;
```

Comment:
I suppose if these were only used for testing, it should be safe to remove…

Reply:
I still need `activityStreamDebug` in here for the purpose of testing though